src/Control/Distributed/Spark/SQL/DataFrame.hs

```diff
@@ -32,3 +33,168 @@ debugDF df = call df "show" []
 
 join :: DataFrame -> DataFrame -> IO DataFrame
 join d1 d2 = call d1 "join" [coerce d2]
+
+newtype DataFrameReader =
+    DataFrameReader (J ('Class "org.apache.spark.sql.DataFrameReader"))
+instance Coercible DataFrameReader
+                   ('Class "org.apache.spark.sql.DataFrameReader")
+
+newtype DataFrameWriter =
+    DataFrameWriter (J ('Class "org.apache.spark.sql.DataFrameWriter"))
+instance Coercible DataFrameWriter
+                   ('Class "org.apache.spark.sql.DataFrameWriter")
+
+read :: SQLContext -> IO DataFrameReader
+read sc = call sc "read" []
+
+write :: DataFrame -> IO DataFrameWriter
+write df = call df "write" []
+
+readParquet :: [Text] -> DataFrameReader -> IO DataFrame
+readParquet fps dfr = do
+    jfps <- reflect fps
+    call dfr "parquet" [coerce jfps]
+
+writeParquet :: Text -> DataFrameWriter -> IO ()
+writeParquet fp dfw = do
+    jfp <- reflect fp
+    call dfw "parquet" [coerce jfp]
+
+newtype StructType =
+    StructType (J ('Class "org.apache.spark.sql.types.StructType"))
+instance Coercible StructType
+                   ('Class "org.apache.spark.sql.types.StructType")
+
+type JStructFieldClass = 'Class "org.apache.spark.sql.types.StructField"
+
+newtype StructField = StructField (J JStructFieldClass)
+instance Coercible StructField
+                   ('Class "org.apache.spark.sql.types.StructField")
+
+schema :: DataFrame -> IO StructType
+schema df = call df "schema" []
+
+fields :: StructType -> IO [StructField]
+fields st = do
+    jfields <- call st "fields" []
+    Prelude.map StructField <$>
+      reify (jfields :: J ('Array JStructFieldClass))
+
+name :: StructField -> IO Text
+name sf = call sf "name" [] >>= reify
+
+nullable :: StructField -> IO Bool
+nullable sf = call sf "nullable" []
+
+newtype DataType = DataType (J ('Class "org.apache.spark.sql.types.DataType"))
+instance Coercible DataType
+                   ('Class "org.apache.spark.sql.types.DataType")
+
+dataType :: StructField -> IO DataType
+dataType sf = call sf "dataType" []
+
+typeName :: DataType -> IO Text
+typeName dt = call dt "typeName" [] >>= reify
+
+newtype Column = Column (J ('Class "org.apache.spark.sql.Column"))
+instance Coercible Column ('Class "org.apache.spark.sql.Column")
+
+type instance Interp Column = 'Class "org.apache.spark.sql.Column"
+instance Reflect Column ('Class "org.apache.spark.sql.Column") where
+  reflect (Column jcol) = return jcol
+
+select :: DataFrame -> [Column] -> IO DataFrame
+select d1 colexprs = do
+  jcols <- reflect colexprs
+  call d1 "select" [coerce jcols]
+
+filterDF :: DataFrame -> Column -> IO DataFrame
```

Comment:
For consistency, either terminate every transformation function with `*DF` or none of them. Otherwise it's more to remember. This module should be imported qualified anyways as it is, since it already takes up high-value variable names. So there's no harm in hiding the Prelude and stealing names from there.